invisible_cities/reco/icaro_components.py

```diff
@@ -0,0 +1,185 @@
+    nbins_e: int
+        Number of bins in energy-axis
+    nsigma_sel: float
+        Number of sigmas to set the band width
+    eff_interval
+        Limits of the range where selection efficiency
+        is considered correct.
+    Returns
+    ----------
+        A  mask corresponding to the selection made.
+    """
+    if input_mask is None:
+        input_mask = [True] * len(dst)
+    else: pass;
+
+    emaps = apply_geo_correction(boot_map, norm_strat  = norm_strat)
+    E0    = dst[input_mask].S2e.values * emaps(dst[input_mask].X.values,
+                                               dst[input_mask].Y.values)
+
+    sel_krband = np.zeros_like(input_mask)
+    sel_krband[input_mask] = selection_in_band(dst[input_mask].Z,
+                                               E0,
+                                               range_z = range_Z,
+                                               range_e = range_E,
+                                               nbins_z = nbins_z,
+                                               nbins_e = nbins_e,
+                                               nsigma  = nsigma_sel)
+
+    effsel   = dst[sel_krband].event.nunique()/dst[input_mask].event.nunique()
+
+    check_if_values_in_interval(data         = np.array(effsel)  ,
+                                minval       = eff_interval[0]   ,
+                                maxval       = eff_interval[1]   ,
+                                display_name = "Z-band selection",
+                                strictness   = strictness        ,
+                                right_closed = True)
+
+    return sel_krband
+
+
+def selection_in_band(z         : np.array,
+                      e         : np.array,
+                      range_z   : Tuple[float, float],
+                      range_e   : Tuple[float, float],
+                      nsigma    : float   = 3.5) ->np.array:
+    """
+    This function returns a mask for the selection of the events that are inside the Kr E vz Z
+
+    Parameters
+    ----------
+    z: np.array
+        axial (z) values
+    e: np.array
+        energy values
+    range_z: Tuple[np.array, np.array]
+        Range in Z-axis
+    range_e: Tuple[np.array, np.array]
+        Range in Energy-axis
+    nsigma: float
+        Number of sigmas to set the band width
+    Returns
+    ----------
+        A  mask corresponding to the selection made.
+    """
+    # Reshapes and flattens are needed for RANSAC function
+
+    z_sel = z[in_range(z, *range_z)]
+    e_sel = e[in_range(e, *range_e)]
+
+    res_fit      = RANSACRegressor().fit(z_sel, np.log(e_sel).reshape(-1, 1))
+
+    in_mask      = res_fit.inlier_mask_
+    residuals_ln = e_sel[in_mask] - np.exp(res_fit.predict(z_sel[in_mask].reshape(-1, 1))).flatten()
+    resy, resx   = np.histogram(residuals_ln, 100)
+    resx         = resx[:-1] + np.diff(resx)
+    fitres       = fit(gauss, resx, resy, seed=[4e3,0,10])
+    fitsigma     = fitres.val[2]
```

Comment:
It might be worth hiding all this logic (sigma estimation) in a separate function to make this function more clear